source/common/router/config_impl.cc

```diff
@@ -337,8 +420,15 @@ const Route* VirtualHostImpl::getRouteFromEntries(const Http::HeaderMap& headers
 
   // Check for a route that matches the request.
   for (const RouteEntryImplBasePtr& route : routes_) {
-    if (route->matches(headers, random_value)) {
-      return route.get();
+    if (route->matches(headers, random_value) != nullptr) {
+      // Cannot push this logic into the RouteEntryImplBase object,
+      // because it leads to leaking the raw pointer to the object,
+      // instead of obtaining the pointer via the shared_ptr.
```

Comment:
per above, we should be using unique_ptr and returning it. This logic should go back into the route implementations. I will comment on that above.